### DISCUSSION TITLE
Create file on USB drive

### DISCUSSION THREAD

**Discussion by jazzjohn27** [earlephilhower/arduino-pico#1006] [Q&A] [answered]
Can a sketch create a file that can be later accessed when the pico w is put into USB drive mode?

**Comment by earlephilhower** [comment, marked as answer]
No, that USB mode is not a real USB stick, only a UF2 bootloader emulation.  There are some examples in TinyUSB of running a MSD (mass storage device/USB drive) mode you can check out, but that's only during normal running mode.

**Reply by zjwhitehead** [reply]
@earlephilhower can you link to a specific example? I saw some activity on the TinyUSB repo around mass storage emulation over usb and would like to use the extra space on my rp2040 internal flash for a file system (trying to write a CSV file from the arduino code and copy it onto a PC via USB)

**Reply by earlephilhower** [reply]
I haven't tried it, but something like https://github.com/adafruit/Adafruit_TinyUSB_Arduino/tree/5be0b21c0a5cf396b139bce0c6c0d403df051b0a/examples/MassStorage/msc_ramdisk

It might also make a lot of sense to just fake a 1-file FAT FS and use the same MSD setup but only have 4-5 sectors of emulated data, like in the bootloader mode.

The onboard filesysrtem is LittleFS, so it's not compatible with PCs, so you can't just export the raw flash as MSD sectors.

**Reply by zjwhitehead** [reply]
Ok, I was hoping to get a version of this working but on the RP2040 **https://github.com/adafruit/Adafruit_TinyUSB_Arduino/blob/5be0b21c0a5cf396b139bce0c6c0d403df051b0a/examples/MassStorage/msc_internal_flash_samd/msc_internal_flash_samd.ino**
I wonder how much work it would be to adapt https://github.com/adafruit/Adafruit_InternalFlash

**Reply by earlephilhower** [reply]
Again, though, if you expose the raw flash as a MSD then the Arduino Core will not be able to read it.  FAT is very bad for flash (writes to the FAT sectors would wear out flash super-fast).

If there was an MTP/PTP mode for TinyUSB that would be the best way.  I don't think that's actually supported, though. :(

**Reply by earlephilhower** [reply]
Check out #1034, @zjwhitehead .  I spent a couple afternoons hacking out a *virtual* FAT stick that actually just reads from the onboard LittleFS.  There's a simple data logger example.  Basically you tell it what LittleFS file will be exported and the `8.3` name and the library handles the magic of pretending there's a FAT drive with that 1 file on it.
